src/runtime/cli/config/configuration-qemu.toml.in

```diff
@@ -16,6 +16,13 @@ kernel = "@KERNELPATH@"
 image = "@IMAGEPATH@"
 machine_type = "@MACHINETYPE@"
 
+# Enable guest protection.
+# Toggling that setting may trigger different hardware features, ranging
+# from memory encryption to both memory and CPU-state encryption and integrity.
+# The Kata Containers runtime dynamically detects the available feature set and
+# aims at enabling the largest possible one.
+guest_protection = false
```

Comment:
I like the idea of having one switch that unites all the different platforms. Most of the time this should be easy. You simply use whatever is supported by the platform you are on. That said, what if I don't want to use all of the available features (i.e. memory encryption but no protection of CPU state). While every platform will support being off or on (with all the features enabled), they might not have the same intermediate values. This probably isn't too big of a deal, but we may want some additional platform specific options field.

What might be more significant is that while it means the same thing for every platform to be off, different platforms provide different guarantees when they are fully enabled. Using one toggle might allow security properties to change silently when workloads are deployed on different nodes. Maybe we should have some concept of minimally acceptable features (enforced by the agent, most likely). One such minimum could be a guarantee that we have any kind of protection enabled on the system. Otherwise it's not totally clear what happens if a user sets guest protection to true but deploys on a node with no security features.

Reply:
Shouldn't this be handled at the Kubernetes layer ?

Reply:
Agreed that if its a heterogeneous cluster, you'd definitely want to have nodes identified before hand in order to take availability into consideration when scheduling. We'd similarly want a way to identify the number of available confidential VMs available (ie, a device plugin), to facilitate scheduling.

Both of these scenarios would be at kubernetes level.

Reply:
Can you trust Kubernetes to schedule confidential workloads on nodes that have the necessary security features? It seems like having appropriate features, or at the very least having some kind of protection, needs to be enforced by something trusted i.e. the agent.

Reply:
QEMU has adopted a slightly different naming. Wonder if we should follow suit.
> A new generic machine option **_confidential-guest-support_** was added to (partially) unify configuration for AMD SEV memory encrypt, POWER PEF and s390 Protected Virtualization, plus future methods of protecting a guest from eavesdropping by a compromised hypervisor.
https://wiki.qemu.org/ChangeLog/6.0

Reply:
@jimcadden `confidential_guest` instead of `guest_protection` ?

Reply:
@jimcadden do you suggest to change it to `confidential_guest_support` ? Is it generic enough to avoid confusions/clashes with other hypervisors (i.e cloud hypervisor) ?

Reply:
@fitzthum The user setting the guest protection to true would eventually translate into selecting a specific runtime class and letting k8s schedule on nodes that are running this runtime class. In an heterogeneous cluster, you could imagine using different runtime classes for different confidential computing implementation/features and exposing (or not, up to the provider to decide) those combinations to the user.

Reply:
> @jimcadden do you suggest to change it to `confidential_guest_support` ? Is it generic enough to avoid confusions/clashes with other hypervisors (i.e cloud hypervisor) ?

I think `confidential_guest` should not be a problem for CLH.

Reply:
I am quite wary of leaving the choice of confidential backend up to the CSP. In the AMD case, SEV, SEV-ES, and SEV-SNP all have different security guarantees. On one node guest protection could mean that the CPU state is protected and the integrity of the memory is protected. On another node guest protection might not provide either of those properties. While Kubernetes will be in charge of selecting nodes for a workload, relying on Kubernetes to provide us a certain set of security features breaks the trust model.

Feature verification should probably happen as part of secret provisioning. With SEV, the guest policy is part of the measurement that is verified prior to secret injection. Thus, the guest owner does not have to trust that Kubernetes chose a suitable node. That said, if the guest owner is only going to inject the secret when certain features are present, Kubernetes will need to be in the loop about which features are required and which nodes have them. I'm not sure how this fits into the vision of a heterogeneous cluster.

Reply:
>   I think confidential_guest should not be a problem for CLH.


:+1: , it sounds good to me

Reply:
@devimc Yes,`confidential_guest` feels slightly more descriptive than `guest_protection`, and it mirrors QEMU's generic terminology for "...methods of protecting a guest from eavesdropping by a compromised hypervisor."  

Also, I'm looking to add `(qemu) confidential_guest_support=` into `govmm` as part of the SEV support here: kata-containers/govmm#167

Reply:
> > I think confidential_guest should not be a problem for CLH.
> 
> +1 , it sounds good to me

Thanks for applying the changes, @devimc

Reply:
> While Kubernetes will be in charge of selecting nodes for a workload, relying on Kubernetes to provide us a certain set of security features breaks the trust model.

If the guest owner is not able to verify the platform its workload is running on top of, then the trust model is broken.
We won't be able to work around the fact that the confidential workload will be scheduled on a node that's selected by an untrusted party (k8s control plane), so we should not rely on k8s to provide us with a certain set of features but only rely on the fact that we should be able to verify if that features set is the one we're expecting. If the k8s operator is not malicious/owned, then this verification step should be following a positive path. If it is malicious, then the worst that can happen is that the confidential workload won't run because it won't be attested and provisioned.

> Feature verification should probably happen as part of secret provisioning. With SEV, the guest policy is part of the measurement that is verified prior to secret injection. Thus, the guest owner does not have to trust that Kubernetes chose a suitable node. Rather, the guest owner can verify exactly which features are enabled prior to injecting the secret. 

The same applies to TDX, which means that if k8s schedules a workload on a node where the measured guest policy does not match the platform feature set, it's the guest owner decision to provision the guest with secrets or not. If it decides that the platform does not provide the right level of confidentiality, it could then reject the provisioning step and the guest would fail to start.

It's not ideal, but given that the node selection is done by an untrusted party, we must protect against that.

> That said, if the guest owner is only going to inject the secret when certain features are present, Kubernetes will need to be in the loop about which features are required and which nodes have them. I'm not sure how this fits into the vision of a heterogeneous cluster.

This certainly makes things difficult and again, bringing kubernetes in the loop does not mean trusting it, the guest owner will still have to verify the platform features, versions, etc before provisioning the guest.

As a tenant/guest owner, if your workload must run on a specific confidential platform, you should have the possibility to select the right runtime class (e.g. `kata-qemu-sev-snp` if the only thing you want to run on is SEV-SNP). Then if k8s schedules the workload accordingly, it should land on a node where this feature set is available and then verified by the guest owner. If that's not the case (cluster misconfiguration, intentional mis-scheduling, etc), then the workload should simply not run on that node.

Does that make sense?